### ISSUE TITLE
Escaping of smart constructors for `@adt` + inconsistency with optics

### ISSUE THREAD

**Issue by kubukoz** [disneystreaming/smithy4s#1347]
1. If you try to generate code for this:

```smithy
@smithy4s.meta#adt
union HasKeywordUnionAdt {
    one: implicit
}

structure implicit {}
```

then you'll get something like this:

```scala
object HasKeywordUnionAdt extends ShapeTag.Companion[HasKeywordUnionAdt] {

  def implicit():Implicit = Implicit()
...
```

and it doesn't compile because of the keyword in the method name.

This is NOT fixed by #1344.

2. I noticed the naming of the smart constructors isn't consistent with the names of optics.

For the example above, we get:

```scala
sealed trait HasKeywordUnionAdt extends scala.Product with scala.Serializable { self =>
  ...
  object project {
    def one: Option[HasKeywordUnionAdt.Implicit] = HasKeywordUnionAdt.Implicit.alt.project.lift(self)
  }
...
}
object HasKeywordUnionAdt extends ShapeTag.Companion[HasKeywordUnionAdt] {

  def implicit():Implicit = Implicit()
  final case class Implicit() extends HasKeywordUnionAdt
...

  trait Visitor[A] {
    def one(value: HasKeywordUnionAdt.Implicit): A
  }

...
}
```

and I don't think that's right.

I see three main options:

- A: move everything to use the target type names in case `UnionMember.ProductCase` is used (`@adt`, `@adtMember`)
- B: rename the smart constructor to match the rest (member name)
- C: escape the smart constructor without renaming and live with the inconsistency.

Personally I'm leaning towards option B, but could be swayed towards A. C sounds like the worst mix.

Here's a WIP for C: https://github.com/disneystreaming/smithy4s/pull/1348

**Comment by Baccata** [comment]
> B: rename the smart constructor to match the rest (member name)

huh, yeah, the smart constructor should be the member name.

**Comment by kubukoz** [comment]
cool, I can get a fix for that soon too then.

**Comment by kubukoz** [comment]
Closed in #1370
